### PR TITLE
cc wrapper: fix a minor problem in cc-wrapper

### PR DESCRIPTION
**Pull request by wrvsrx**
**This comment is outdated, check https://github.com/NixOS/nixpkgs/pull/169327#issuecomment-1409711315 for latest description of this PR.**

###### Description of changes

I made two minor changes to `cc-wrapper`

1.  When `cc-wrapper` tests languages, it only tests `-x c++` flag but ignoring the `-xc++` flag. I add the case.
2.  `clang` provides a flag `-cxx-isystem` to specify headers only for c++ (It works for all `llvm` version in nixpkgs. These paths are ignored for c program). Therefore we don't need to manual test for language then adding `-isystem` flag in wrapper, just use `-cxx-isystem`.
    
    This change fixes a case. When compiling a cpp file using `clang` such as `clang test.cpp -stdlib=stdc++ -lstdc++`, normal clang succeed but this wrapper fails. Although `clang` can recognize it's cpp, but this wrapper doesn't add `-isystem <path to c++ header>` since the wrapper doesn't detect the `-x c++` flag.

Besides, I thick some further changes should be made. In other linux system (such as arch linux), libraries such as `-lc++`, `-lc++abi` needed be added manually when using `clang` instead of `clang++` to compile c++ programs, even if `-x c++` is specified. But in this wrapper, these flags are added automatically when using `clang -x c++`. To have consistent behavior with normal clang (I mean clang built from source), additional testing and flag adding for c++ should be removed. If you maintainers think that makes sense, please comment under this PR. Then I'll make a PR to do the further change.

ref: https://releases.llvm.org/13.0.0/tools/clang/docs/ClangCommandLineReference.html

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by acowley** [comment]
I am not a maintainer, but I would definitely like to simplify the language detection situation. The differences between the nixpkgs clang wrappers and other deployment scenarios do cause me some grief.

**Comment by Artturin** [comment]
should target staging https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#rebasing-between-branches-ie-from-master-to-staging

**Comment by wrvsrx** [comment]
Opps, I do the instruction in a wrong order, I'll close the PR and make the PR again.

**Comment by RaitoBezarius** [comment]
Could we get a review on this PR @Ericson2314 or should we do an open call for someone to review this?
Also, @wrvsrx would you have time to fix the conflicts in this PR?

**Comment by wrvsrx** [comment]
@RaitoBezarius Done.

File: pkgs/build-support/cc-wrapper/default.nix [hunk 1]
```diff
@@ -359,14 +359,14 @@ stdenv.mkDerivation {
     ''
     + optionalString (libcxx == null && (useGccForLibs && gccForLibs.langCC or false)) ''
       for dir in ${gccForLibs}${lib.optionalString (hostPlatform != targetPlatform) "/${targetPlatform.config}"}/include/c++/*; do
-        echo "-isystem $dir" >> $out/nix-support/libcxx-cxxflags
+        echo "-cxx-isystem $dir" >> $out/nix-support/libc-cflags
       done
       for dir in ${gccForLibs}${lib.optionalString (hostPlatform != targetPlatform) "/${targetPlatform.config}"}/include/c++/*/${targetPlatform.config}; do
-        echo "-isystem $dir" >> $out/nix-support/libcxx-cxxflags
```
Comment (Mic92) [inline pkgs/build-support/cc-wrapper/default.nix:365]: Note that these flags are not just consumed by clang but also by some other tools i.e. bindgen for rust.
I think most of these third-party tools are llvm-based but we should at least check if they also support this prefix:

- bindgen
- ccls
- clang-tools/clangd (very likely)

Comment (wrvsrx) [reply pkgs/build-support/cc-wrapper/default.nix:365]: You're right. I'll add a commit later to add some patches for tools using libcxx-cxxflags.

**Comment by wrvsrx** [comment]
@Mic92 

I made two changes after your last review.

1. I changed clang-tools' wrapper to make it work with `-cxx-isystem`.
2. I added `-cxx-isystem` to `libcxx-cxxflags` instead of `libc-clags`. I said "we don't need to manual test for language then adding -isystem flag in wrapper, just use -cxx-isystem" in original PR, it' wrong. Language detection in wrapper is necesarry if we want `-nostdinc++` and `-nostdinc` flags work properly, therefore `-cxx-isystem` should be added to `libcxx-cxxflags`.

Now I have tested rust-bindgen, clang-tools and ccls: I test rust-bingen by using it generate a hpp file containing iostream, test clang-tools and ccls by using them as language server in editor (neovim) and test if they can find c++ header correctly. All of them work well. 

In summary, this PR made two changes:

1. When cc-wrapper tests languages, it only tests -x c++ flag but ignoring the -xc++ flag. I add the case.
2. change `-isystem` flag for c++ header to `-cxx-isystem`.

**Review by Mic92 — APPROVED** [review]
Changes look solid. I have yet been able to test it myself.

**Comment by RaitoBezarius** [comment]
> Changes look solid. I have yet been able to test it myself.

How would you test it? (for reference)

**Comment by wrvsrx** [comment]
Fix merge conflicts.

**Comment by wrvsrx** [comment]
Revert change to `-cxx-isystem`. Because there's a case: If we use gcc and libc++, then `-cxx-isystem` will be passed to gcc, and gcc doesn't support that flag. Thanks for the advice from @NickCao .

Now this PR only make one change:

- When cc-wrapper tests languages, it only tests `-x c++` flag but ignoring the `-xc++` flag (It also ignores `-x*-header`). I add the case.

**Review by NickCao — APPROVED** [review]

**Comment by RaitoBezarius** [comment]
I think we have 2 approvals, no one would oppose to merge, right? (will merge on Tuesday otherwise.)
